animalcontrol.lua

```diff
@@ -0,0 +1,211 @@
+ 'markfor',
+ 'unmarkfor',
+
+ 'help',
+})
+local args = utils.processArgs({...}, validArgs)
+local help = [====[
+
+animalcontrol
+=============
+Animal control is a script useful for deciding what animals to butcher and geld.
+
+I developed this because I found Dwarf Therapist had a bug for my version of
+dwarf fortress which prevented the geld status of animals from updating. So
+I would be queuing animals to be gelded that had already been gelded.
+
+While not as powerful as Dwarf Therapist in managing animals - in so far as
+DT allows you to sort by various stats and flags - this script does provide
+many options for filtering animals. Additionally you can mark animals for
+slaughter or gelding, you can even do so enmasse if you so choose.
+
+Examples:
+  [DFHack]# animalcontrol -race 170
+  [DFHack]# animalcontrol -race 170 -male -notgelded -showstats
+  [DFHack]# animalcontrol -markfor gelding -id 1988
+  [DFHack]# animalcontrol -markfor slaughter -id 1988
+  [DFHack]# animalcontrol -gelded -markedfor slaughter -unmarkfor slaughter
+
+-~~~~~~~~~~~
+selection options:
+
+  These options are used to specify what animals you want or do not want to select.
+    all                  - Selects all units
+                           Note: cannot be used in conjunction with other selection options.
+
+    id <value>           - Selects the unit with the specified id value provided.
+    race <value>         - Selects units which match the race value provided.
+
+    markedfor <value>    - Selects units which have been marked for the value provided.
+                           valid values: slaughter, gelding
+
+    notmarkedfor <value> - Selects units which have not been marked for the value provided.
+                           valid values: slaughter, gelding
+
+    gelded               - Selects units which have already been gelded.
+    notgelded            - Selects units which have not been gelded.
+    male                 - Selects units which are male.
+    female               - Selects units which are female.
+
+-~~~~~~~~~~~
+command options:
+
+    showstats            - Displays physical attributes of the selected animals.
+
+    markfor <value>      - Marks selected animals for the value provided.
+                           valid values: slaughter, gelding
+
+    unmarkfor <value>    - Unmarks selected animals for the value provided.
+                           valid values: slaughter, gelding
+
+other options:
+    help                 - Displays this information
+
+]====]
+
+local Units = df.global.world.units.active
+
+if args.race and not tonumber(args.race) then
+    args.race=string.upper(args.race)
+    local raceID
+    for i,c in ipairs(df.global.world.raws.creatures.all) do
+      if c.creature_id == args.race then
+        raceID = i
+        break
+      end
+    end
+    if not raceID then
+      qerror('Invalid race: ' .. args.race)
+    end
+    args.race = raceID
+end
+
+function safe_pairs(item, keys_only)
```

Comment:
Is this copied from another script? (It's fine if it is, I just might move it to the utils module if it's useful.)

Reply:
The safe pairs I believe is. I think I found it in a file you wrote back in 2017, I use it in a lot of the scripts I work on.

Reply:
Looks like regular pairs() or ipairs() should be fine here, since you only use it on world.units.active, which always supports those functions